### PR TITLE
docker trust signer-remove

### PR DESCRIPTION
**Pull request by eiais**
Signed-off-by: Kyle Spiers <kyle@spiers.me>

### CODE REVIEW

**Comment by ashfall** [comment]
We don't have `-` in our filenames much, can we switch `cli/command/trust/signer-remove.go` to `cli/command/trust/signer_remove.go` ?

**Review by ashfall — COMMENTED** [review]
We don't have `-` in our filenames much, can we switch `cli/command/trust/signer-remove.go` to `cli/command/trust/signer_remove.go` ?

Update: Looks like codecov doesn't recognize `_` files well. This is probably fine as is.

File: cli/command/trust/signer-remove.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/docker/cli/cli/trust"
+	//"github.com/docker/notary"
+	//"github.com/docker/notary/client"
+	"github.com/docker/notary/tuf/data"
+	"github.com/spf13/cobra"
+)
+
+func newSignerRemoveCommand(dockerCli command.Cli) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "signer-remove IMAGE NAME",
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: we should make this command work across multiple images. Perhaps it changes to:

`signer-remove [OPTIONS] NAME IMAGE [IMAGE...]` ?

File: cli/command/trust/signer-remove.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		//return some error
+	}
+	err = notaryRepo.RemoveDelegationRole(data.RoleName(name))
+	if err != nil{
+		return err
+	}
+	err = notaryRepo.Publish()
+	if err != nil{
+		return err
+	}
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: from feedback: let's have a success message here

File: cli/command/trust/signer-remove.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	//"github.com/docker/notary"
+	//"github.com/docker/notary/client"
+	"github.com/docker/notary/tuf/data"
+	"github.com/spf13/cobra"
+)
+
+func newSignerRemoveCommand(dockerCli command.Cli) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "signer-remove IMAGE NAME",
+		Short: "Display detailed information about keys and signatures",
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: need to update to reflect removing a signer

File: cli/command/trust/signer-remove.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		Short: "Remove a signer",
+		Args:  cli.RequiresMinArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			options.images = args[1:]
+			return removeSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	flags.BoolVarP(&options.forceYes, "yes", "y", false, "Answer yes to removing most recent signer (no confirmation)")
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: can we use this function to keep consistent with `docker/cli`?
  https://github.com/docker/cli/pull/472/files#diff-7054cfd4e7f081e2955063c1bd7da5e6R45

(I learned about this func from code review)

File: cli/command/trust/signer-remove.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+type signerRemoveOptions struct {
+	forceYes bool
+	images   []string
+	signer   string
+}
+
+func newSignerRemoveCommand(dockerCli command.Cli) *cobra.Command {
+	options := signerRemoveOptions{}
+	cmd := &cobra.Command{
+		Use:   "signer-remove IMAGE NAME",
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: `signer-remove NAME IMAGE [IMAGE...]`

File: cli/command/trust/signer-remove.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		},
+	}
+	flags := cmd.Flags()
+	flags.BoolVarP(&options.forceYes, "yes", "y", false, "Answer yes to removing most recent signer (no confirmation)")
+	return cmd
+}
+
+func removeSigner(cli command.Cli, options *signerRemoveOptions) error {
+	if len(options.images) < 1 {
+		return fmt.Errorf("at least one signer must be provided")
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: s/signer/image

File: cli/command/trust/signer-remove.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+		return true, err
+	}
+	var lastSigner client.RoleWithSignatures
+	for i := range allRoles {
+		if allRoles[i].Name == "targets/releases" {
+			lastSigner = allRoles[i]
+		}
+	}
+	if len(lastSigner.Signatures) == 0 {
+		fmt.Printf("All tags are currently unpullable, use docker sign to fix")
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: s/docker sign/docker trust sign

File: cli/command/trust/signer-remove.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+		return trust.NotaryError(ref.Name(), err)
+	}
+	delegation := "targets/" + name
+	var role data.Role
+	if delegation == "targets/releases" {
+		//return some error
+	}
+	delegationRoles, err := notaryRepo.GetDelegationRoles()
+	if err != nil {
+		return fmt.Errorf("Error retrieving signers for %s: %s", name, image)
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: I don't think we need to display the signer name here, can just be image in the err

File: cli/command/trust/signer-remove.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	for i := range delegationRoles {
+		if delegationRoles[i].Name == data.RoleName(delegation) {
+			role = delegationRoles[i]
+		}
+	}
+	if role.Name == "" {
+		return fmt.Errorf("No  signer %s for image %s", image, err)
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: I think this should be `fmt.Errorf("No signer %s for image %s", name, image)`

File: cli/command/trust/signer-remove.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+	images   []string
+	signer   string
+}
+
+func newSignerRemoveCommand(dockerCli command.Cli) *cobra.Command {
+	options := signerRemoveOptions{}
+	cmd := &cobra.Command{
+		Use:   "signer-remove IMAGE NAME",
+		Short: "Remove a signer",
+		Args:  cli.RequiresMinArgs(0),
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: Should this be `2`?

File: cli/command/trust/signer-remove.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+	for i := range allRoles {
+		if allRoles[i].Name == "targets/releases" {
+			lastSigner = allRoles[i]
+		}
+	}
+	if len(lastSigner.Signatures) == 0 {
+		fmt.Printf("All tags are currently unpullable, use docker sign to fix")
+		return false, nil
+	}
+	return (role.KeyIDs[0] == lastSigner.Signatures[0].KeyID), nil
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: While it may not be possible to produce this metadata with a notary client today, I think we should do a double for loop over role key ids and signature key ids.

Also, it will prevent us from indexing into a potentially empty slice.

File: cli/command/trust/signer-remove.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+func removeSingleSigner(cli command.Cli, image, name string, forceYes bool) error {
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: nit: can we change `name` to `signerName`?

File: cli/command/trust/signer-remove.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+		return err
+	}
+	notaryRepo, err := trust.GetNotaryRepository(cli, repoInfo, *authConfig, "push", "pull")
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+	delegation := "targets/" + name
+	var role data.Role
+	if delegation == "targets/releases" {
+		//return some error
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: @ashfall has a similar case in `signer-add` with a good current error message.  We should keep consistent between the two subcommands, could you please coordinate with her?

File: cli/command/trust/signer-remove.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+		removeSigner := askConfirm(cli.In())
+		if !removeSigner {
+			fmt.Fprintf(cli.Out(), "\nAborting action. \n")
+			return nil
+		}
+	} else if err != nil {
+		return err
+	}
+	keys := role.KeyIDs
+	if err = notaryRepo.RemoveDelegationKeys("targets/releases", keys); err != nil {
```
Comment (riyazdf) [inline cli/command/trust/signer-remove.go]: do we have a const for `targets/releases`?  If not, could we make one?

**Review by riyazdf — COMMENTED** [review]
this is shaping up nicely, looking forward to tests!

File: cli/command/trust/signer_remove.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+		return true, err
+	}
+	var lastSigner client.RoleWithSignatures
+	for i := range allRoles {
+		if allRoles[i].Name == "targets/releases" {
+			lastSigner = allRoles[i]
+		}
+	}
+	if len(lastSigner.Signatures) == 0 {
+		fmt.Printf("All tags are currently unpullable, use docker trust sign to fix")
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: maybe we can reword this to highlight that it only affects signed tags, and that they've been revoked in some form?  Maybe: `All signed tags are currently revoked, use docker trust sign to fix` ?

File: cli/command/trust/signer_remove.go [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+func isLastSigner(role data.Role, notaryRepo *client.NotaryRepository) (bool, error) {
+	allRoles, err := notaryRepo.ListRoles()
+	if err != nil {
+		return true, err
+	}
+	var lastSigner client.RoleWithSignatures
+	for i := range allRoles {
+		if allRoles[i].Name == "targets/releases" {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: use const `releases`

File: cli/command/trust/signer_remove.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+			lastSigner = allRoles[i]
+		}
+	}
+	if len(lastSigner.Signatures) == 0 {
+		fmt.Printf("All tags are currently unpullable, use docker trust sign to fix")
+		return false, nil
+	}
+	nonRoleKey := false
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: let's rename this to `hasSignedReleases`?

File: cli/command/trust/signer_remove.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+	for _, image := range options.images {
+		if err := removeSingleSigner(cli, image, options.signer, options.forceYes); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isLastSigner(role data.Role, notaryRepo *client.NotaryRepository) (bool, error) {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: it'll be easier to test this function if it takes the output of `notaryRepo.ListRoles()` as an arg instead of a direct `NotaryRepository`

File: cli/command/trust/signer_remove.go [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+	delegation := "targets/" + signerName
+	var role data.Role
+	if delegation == releases {
+		//return some error
+	}
+	delegationRoles, err := notaryRepo.GetDelegationRoles()
+	if err != nil {
+		return fmt.Errorf("Error retrieving signers for %s", image)
+	}
+	for i := range delegationRoles {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: slightly more golang-esque:

```
for _, delRole := range delegationRoles {
  if delRole.Name == ...
}

File: cli/command/trust/signer_remove.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+		return err
+	}
+	notaryRepo, err := trust.GetNotaryRepository(cli, repoInfo, *authConfig, "push", "pull")
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+	delegation := "targets/" + signerName
+	var role data.Role
+	if delegation == releases {
+		//return some error
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: we should mirror the error on `signer-add`.  Best way to do this is probably to create a new error type and share them between the two commands

File: cli/command/trust/signer_remove.go [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+			role = delegationRoles[i]
+		}
+	}
+	if role.Name == "" {
+		return fmt.Errorf("No  signer %s for image %s", signerName, image)
+	}
+	if ok, err := isLastSigner(role, notaryRepo); ok && !forceYes && err == nil {
+		removeSigner := command.PromptForConfirmation(os.Stdin, cli.Out(), fmt.Sprintf("The signer %s signed the last released version of %s. "+
+			"Removing them will lead to %s being unpullable."+
+			"Are you sure you want to continue? (y/n): ",
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: no need for `(y/n)` here, we'll get `[y/N]` from the prompt for free

File: cli/command/trust/signer_remove.go [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+		))
+
+		if !removeSigner {
+			fmt.Fprintf(cli.Out(), "\nAborting action. \n")
+			return nil
+		}
+	} else if err != nil {
+		return err
+	}
+	keys := role.KeyIDs
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: we can inline this into the function call below

File: cli/command/trust/signer_remove.go [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isLastSigner(role data.Role, allRoles []client.RoleWithSignatures) bool {
+	var lastSigner client.RoleWithSignatures
+	for i := range allRoles {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: slightly more idiomatic:

```
for _, role := range allRoles {
...
}
```

File: cli/command/trust/signer_remove.go [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+		if allRoles[i].Name == "targets/releases" {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: const: `releases` (edit: or `releasesRoleTUFName` if renamed)

File: cli/command/trust/helpers.go [hunk 25]
```diff
@@ -17,6 +17,7 @@ import (
 )
 
 const releasedRoleName = "Repo Admin"
+const releases = "targets/releases"
```
Comment (riyazdf) [inline cli/command/trust/helpers.go]: can this be: `releasesRoleTUFName` ?

File: cli/command/trust/signer_remove.go [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	return true
+}
+
+func removeSingleSigner(cli command.Cli, image, signerName string, forceYes bool) error {
+	_, ref, repoInfo, authConfig, err := getImageReferencesAndAuth(cli, image)
+	if err != nil {
+		return err
+	}
+	delegation := "targets/" + signerName
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: can we rename this to `signerDelegation` for clarity?

File: cli/command/trust/signer_remove.go [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return fmt.Errorf("Error retrieving signers for %s", image)
+	}
+	for _, delRole := range delegationRoles {
+		if delRole.Name == data.RoleName(delegation) {
+			role = delRole
+		}
+	}
+	if role.Name == "" {
+		return fmt.Errorf("No  signer %s for image %s", signerName, image)
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: extra space after `No`

File: cli/command/trust/signer_remove.go [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	allRoles, err := notaryRepo.ListRoles()
+	if err != nil {
+		return err
+	}
+	if ok := isLastSigner(role, allRoles); ok && !forceYes {
+		removeSigner := command.PromptForConfirmation(os.Stdin, cli.Out(), fmt.Sprintf("The signer %s signed the last released version of %s. "+
+			"Removing them will lead to %s being unpullable."+
+			"Are you sure you want to continue? (y/n): ",
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: I think the `(y/n)` is unnecessary, it should be provided by `command.PromptForConfirmation`

File: cli/command/trust/signer_remove.go [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+		))
+
+		if !removeSigner {
+			fmt.Fprintf(cli.Out(), "\nAborting action. \n")
+			return nil
+		}
+	} else if err != nil {
+		return err
+	}
+	if err = notaryRepo.RemoveDelegationKeys(releases, role.KeyIDs); err != nil {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: can we have a test for what happens when the signer's key is not in `targets/releases`, but they have their own role still?

Also: if the signer is neither in `targets/releases` nor `targets/<signer>`?

I'm not sure if we can do much here without a manual test for now.  It would be good to track these tests in the existing issue or a new one.

File: cli/command/trust/signer_remove.go [hunk 30]
```diff
@@ -0,0 +1,129 @@
+package trust
+
+import (
+	"fmt"
+
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: nit: remove extra line

**Review by riyazdf — COMMENTED** [review]
Looking good, almost there!  Would be great to get some example CLI output and handle a couple more test cases to the best we can

**Comment by riyazdf** [comment]
looks like your test also needs to remove an extra space character (the `received` is correct):
```
--- FAIL: TestRemoveSingleSigner (0.52s)
	Error Trace:	signer_remove_test.go:57
	Error:      	Error message not equal:
	            	expected: "No  signer test for image eiais/test2"
	            	received: "No signer test for image eiais/test2"
```

File: cli/command/trust/signer_remove.go [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+	delegationRoles, err := notaryRepo.GetDelegationRoles()
+	if err != nil {
+		return fmt.Errorf("Error retrieving signers for %s", image)
+	}
+	for _, delRole := range delegationRoles {
+		if delRole.Name == data.RoleName(signerDelegation) {
+			role = delRole
```
Comment (ashfall) [inline cli/command/trust/signer_remove.go]: Should we break out of the loop once a matching delegation role is found?

File: cli/command/trust/signer_remove.go [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	if ok := isLastSigner(role, allRoles); ok && !forceYes {
+		removeSigner := command.PromptForConfirmation(os.Stdin, cli.Out(), fmt.Sprintf("The signer %s signed the last released version of %s. "+
+			"Removing them will lead to %s being unpullable."+
+			"Are you sure you want to continue?: ",
+			signerName, image, image,
+		))
+
+		if !removeSigner {
+			fmt.Fprintf(cli.Out(), "\nAborting action. \n")
```
Comment (ashfall) [inline cli/command/trust/signer_remove.go]: nit: Do we need the space before the `\n`?

File: cli/command/trust/signer_remove.go [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+			return nil
+		}
+	} else if err != nil {
```
Comment (ashfall) [inline cli/command/trust/signer_remove.go]: Where is this `err` originating?

File: cli/command/trust/signer_remove.go [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+			return removeSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	flags.BoolVarP(&options.forceYes, "yes", "y", false, "Answer yes to removing most recent signer (no confirmation)")
+	return cmd
+}
+
+func removeSigner(cli command.Cli, options *signerRemoveOptions) error {
+	if len(options.images) < 1 {
```
Comment (ashfall) [inline cli/command/trust/signer_remove.go]: Will this case ever be true? I think it'll be caught in `newSignerRemoveCommand`'s `RequiresMinArgs(2)` instead, and this check is unnecessary.

File: cli/command/trust/signer_remove.go [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+
+func isLastSigner(role data.Role, allRoles []client.RoleWithSignatures) bool {
+	var lastSigner client.RoleWithSignatures
+	for _, role := range allRoles {
+		if role.Name == releasesRoleTUFName {
+			lastSigner = role
+		}
+	}
+	if len(lastSigner.Signatures) == 0 {
+		fmt.Printf("All signed tags are currently revoked, use docker trust sign to fix")
```
Comment (ashfall) [inline cli/command/trust/signer_remove.go]: I think we should use `fmt.Fprintf` here. Though that would require the `cli` to be passed in here. :(

If this print is necessary, then perhaps we can refactor the actual logic out and print this someplace that already has the `cli`. Thoughts?

**Review by ashfall — COMMENTED** [review]
Good work, thanks for adding this!

File: cli/command/trust/signer_remove.go [hunk 36]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	return nil
+}
+
+func isLastSigner(role data.Role, allRoles []client.RoleWithSignatures) (bool, error) {
+	var lastSigner client.RoleWithSignatures
+	for _, role := range allRoles {
+		if role.Name == releasesRoleTUFName {
+			lastSigner = role
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: we can add a `break` after this statement, and also rename `lastSigner` to `releasesRoleWithSigs` for clarity?

File: cli/command/trust/signer_remove.go [hunk 37]
```diff
@@ -0,0 +1,124 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isLastSigner(role data.Role, allRoles []client.RoleWithSignatures) (bool, error) {
+	var lastSigner client.RoleWithSignatures
+	for _, role := range allRoles {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: can we change `role` here to `roleWithSig` to not overload the `role` arg in this block?

File: cli/command/trust/signer_remove.go [hunk 38]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/trust"
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/tuf/data"
+	"github.com/spf13/cobra"
+)
+
+type signerRemoveOptions struct {
+	forceYes bool
+	images   []string
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: we shouldn't need `images` or `signer` in this options block, I would prefer to pass`args[0]` and `args[1:]` directly to `removeSigner`

File: cli/command/trust/signer_remove.go [hunk 39]
```diff
@@ -0,0 +1,124 @@
...
+	return true, nil
+}
+
+func removeSingleSigner(cli command.Cli, image, signerName string, forceYes bool) error {
+	_, ref, repoInfo, authConfig, err := getImageReferencesAndAuth(cli, image)
+	if err != nil {
+		return err
+	}
+	signerDelegation := "targets/" + signerName
+	var role data.Role
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: let's move this var declaration directly above where it's needed

File: cli/command/trust/signer_remove.go [hunk 40]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	if role.Name == "" {
+		return fmt.Errorf("No signer %s for image %s", signerName, image)
+	}
+	allRoles, err := notaryRepo.ListRoles()
+	if err != nil {
+		return err
+	}
+	if ok, err := isLastSigner(role, allRoles); ok && !forceYes {
+		removeSigner := command.PromptForConfirmation(os.Stdin, cli.Out(), fmt.Sprintf("The signer %s signed the last released version of %s. "+
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: this message should clarify that this only pertains to signed tags.  Maybe "...Removing them will lead to signed tags for %s to be unpullable" ?

File: cli/command/trust/signer_remove.go [hunk 41]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	return true, nil
+}
+
+func removeSingleSigner(cli command.Cli, image, signerName string, forceYes bool) error {
+	_, ref, repoInfo, authConfig, err := getImageReferencesAndAuth(cli, image)
+	if err != nil {
+		return err
+	}
+	signerDelegation := "targets/" + signerName
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: let's make this a `data.RoleName` type here, since we cast more often than use it directly as a string

File: cli/command/trust/signer_remove.go [hunk 42]
```diff
@@ -0,0 +1,124 @@
...
+	for _, image := range options.images {
+		if err := removeSingleSigner(cli, image, options.signer, options.forceYes); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isLastSigner(role data.Role, allRoles []client.RoleWithSignatures) (bool, error) {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: can we rename this to `isLastSignerForReleases` ?

File: cli/command/trust/signer_remove.go [hunk 43]
```diff
@@ -0,0 +1,124 @@
...
+		if role.Name == releasesRoleTUFName {
+			lastSigner = role
+		}
+	}
+	if len(lastSigner.Signatures) == 0 {
+		return false, fmt.Errorf("All signed tags are currently revoked, use docker trust sign to fix")
+	}
+	for _, signature := range lastSigner.Signatures {
+		for _, key := range role.KeyIDs {
+			if signature.KeyID != key {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: I think this should instead check `if signature.KeyID == key` and technically also check against the threshold?  We should count the number of keys from the role that have signed the  threshold and `return false, error` if `len(Signatures) - count < threshold`

File: cli/command/trust/signer_remove.go [hunk 44]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	}
+	if role.Name == "" {
+		return fmt.Errorf("No signer %s for image %s", signerName, image)
+	}
+	allRoles, err := notaryRepo.ListRoles()
+	if err != nil {
+		return err
+	}
+	if ok, err := isLastSigner(role, allRoles); ok && !forceYes {
```
Comment (riyazdf) [inline cli/command/trust/signer_remove.go]: let's also check `err != nil` to be defensive

File: cli/command/trust/signer_remove.go (hunk 43 above)
Comment (eiais) [reply cli/command/trust/signer_remove.go]: I think ```signature.KeyID != key``` is correct. Wouldn't this require knowing ucp signing policy?

File: cli/command/trust/signer_remove.go (hunk 44 above)
Comment (eiais) [reply cli/command/trust/signer_remove.go]: I am

File: cli/command/trust/signer_remove.go (hunk 43 above)
Comment (riyazdf) [reply cli/command/trust/signer_remove.go]: from IRL discussion: as @eiais mentions, we should increment on `signature.KeyID != key`.  This is TUF thresholding separate from UCP signing policy

**Review by riyazdf — APPROVED** [review]
LGTM, cc @ashfall for final review before merge

File: cli/command/trust/sign.go [hunk 45]
```diff
@@ -64,7 +64,7 @@ func signImage(cli command.Cli, imageName string) error {
 			}
 
 			fmt.Fprintf(cli.Out(), "Created signer: %s\n", authConfig.Username)
-			fmt.Fprintf(cli.Out(), "Finished initializing %q\n", notaryRepo.GetGUN().String())
+			fmt.Fprintf(cli.Out(), "Finished initializing signed repository for %s\n", imageName)
```
Comment (ashfall) [inline cli/command/trust/sign.go]: minor nit/note: I think this change is already in master (https://github.com/riyazdf/cli/blob/master/cli/command/trust/sign.go#L67), the diff coverage should go up if you rebase, and will keep the diff's scope smaller.

**Review by ashfall — APPROVED** [review]
LGTM! One comment about rebasing, but non-blocking. Feel free to merge!

**Review by riyazdf — APPROVED** [review]
LGTM, merge on green!
